network/transport/v2/retry.go

```diff
@@ -0,0 +1,195 @@
+type Retriable interface {
+	// Add a job that needs to be retried. It'll be passed to the channel immediately and at set intervals.
+	// Returns nil if job already exists.
+	// An error is returned if there's a problem with the underlying storage.
+	Add(hash hash.SHA256Hash) error
+	// Remove a job, it'll no longer be retried.
+	// An error is returned if there's a problem with the underlying storage.
+	Remove(hash hash.SHA256Hash) error
+	// Configure opens the DB connection for persistent job storage and loads existing jobs from disk
+	Configure() error
+	// Start retrying new and existing jobs
+	Start() error
+	// Close cancels all jobs and closes the DB
+	Close() error
+}
+
+type jobCallBack func(hash hash.SHA256Hash)
+
+// NewPayloadRetrier returns a Retriable for payload fetches.
+// The payload hashes as []byte should be added as job.
+func NewPayloadRetrier(config Config, callback jobCallBack) Retriable {
+	return &payloadRetrier{
+		config:   config,
+		callback: callback,
+	}
+}
+
+type payloadRetrier struct {
+	config   Config
+	db       *bbolt.DB
+	callback jobCallBack
+	ctx      context.Context
+	cancel   context.CancelFunc
+}
+
+func (p *payloadRetrier) Configure() error {
+	p.ctx, p.cancel = context.WithCancel(context.Background())
+
+	dbFile := path.Join(p.config.Datadir, "network", "payload_jobs.db")
+	if err := os.MkdirAll(filepath.Dir(dbFile), os.ModePerm); err != nil {
+		return fmt.Errorf("unable to setup database: %w", err)
+	}
+
+	if p.config.PayloadRetryDelay == 0 {
+		p.config.PayloadRetryDelay = 5 * time.Second
+	}
+	var err error
+	p.db, err = bbolt.Open(dbFile, 0600, bbolt.DefaultOptions)
+	if err != nil {
+		return fmt.Errorf("unable to create BBolt database: %w", err)
+	}
+	if err = p.db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists([]byte("payload_jobs"))
+		return err
+	}); err != nil {
+		return fmt.Errorf("unable to create buckets in database: %w", err)
+	}
+	return nil
+}
+
+func (p *payloadRetrier) Start() error {
+	return p.db.View(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket([]byte("payload_jobs"))
+		return bucket.ForEach(func(k, v []byte) error {
+			h := hash.FromSlice(k)
+			c := binary.LittleEndian.Uint16(v)
+			p.retry(h, c)
+			return nil
+		})
+	})
+}
+
+func (p *payloadRetrier) Add(hash hash.SHA256Hash) error {
+	if err := p.writeCount(hash, 0); err != nil {
+		return err
+	}
+	p.retry(hash, 0)
+
+	return nil
+}
+
+var errContinue = errors.New("continue")
+
+func (p *payloadRetrier) retry(hash hash.SHA256Hash, initialCount uint16) {
+	startDelay := p.config.PayloadRetryDelay
+	for i := uint16(0); i < initialCount; i++ {
+		startDelay *= 2
+	}
+
+	go func() {
+		err := retry.Do(func() error {
+			count, existing, err := p.readCount(hash)
+			if err != nil {
+				return retry.Unrecoverable(err)
+			}
+			if existing {
+				if err := p.writeCount(hash, count+1); err != nil {
+					return retry.Unrecoverable(err)
+				}
+				p.callback(hash)
+				return errContinue
```

Comment:
what if the callback causes the payload to be retrieved successfully, is an error logged by the retry lib? (because we're returning an error here)